src/Pate/Verification/StrongestPosts.hs

```diff
@@ -0,0 +1,917 @@
+   -- using the ordinary widening machinery. However, it seems like there might be a bug in this
+   -- procedure, as we are currently get a lot of widening errors that seem due to the propagation
+   -- along return vectors.
+
+   processReturn gr ret = withPair ret $
+     do (asm, bundle) <- returnSiteBundle d ret
+        traceBundle bundle "Processing return edge"
+--        traceBundle bundle "== bundle asm =="
+--        traceBundle bundle (show (W4.printSymExpr asm))
+        withEmptyAssumptionFrame $
+          withAssumption_ (return asm) $
+          widenAlongEdge bundle (ReturnNode fPair) d gr (GraphNode ret)
+
+
+-- | Construct a "dummy" simulation bundle that basically just
+--   immediately returns the prestate as the poststate.
+--   This is used to compute widenings that propagate abstract domain
+--   information from function return nodes to the actual return sites.
+--
+--   TODO, I'm not convinced this is working correctly.
+returnSiteBundle :: forall sym arch.
+  AbstractDomain sym arch ->
+  PPa.BlockPair arch {- ^ block pair being returned to -} ->
+  EquivM sym arch (W4.Pred sym, SimBundle sym arch)
+returnSiteBundle preD pPair =
+  withEmptyAssumptionFrame $
+  withSym $ \sym ->
+  do let oVarState = PS.simVarState (PPa.pOriginal (PS.specVars preD))
+     let pVarState = PS.simVarState (PPa.pPatched  (PS.specVars preD))
+
+     let simInO    = PS.SimInput oVarState (PPa.pOriginal pPair)
+     let simInP    = PS.SimInput pVarState (PPa.pPatched pPair)
+
+     blockEndVal <- liftIO (MS.initBlockEnd (Proxy @arch) sym)
+
+     let simOutO   = PS.SimOutput oVarState blockEndVal
+     let simOutP   = PS.SimOutput pVarState blockEndVal
+
+     -- TODO? Not clear how much this rewrite step is going to do for us.  Perhaps just skip it?
+     withAssumptionFrame (PVV.validInitState (Just pPair) oVarState pVarState) $
+       applyCurrentFrame
+         (SimBundle (PPa.PatchPair simInO simInP)
+                    (PPa.PatchPair simOutO simOutP))
+
+data ObservableCheckResult sym ptrW
+  = ObservableCheckEq
+  | ObservableCheckCounterexample (ObservableCounterexample sym ptrW)
+  | ObservableCheckError String
+
+checkObservables :: forall sym arch.
+  PPa.BlockPair arch ->
+  W4.Pred sym ->
+  SimBundle sym arch ->
+  AbstractDomain sym arch ->
+  PairGraph sym arch ->
+  EquivM sym arch (PairGraph sym arch)
+checkObservables bPair asm bundle preD gr =
+  case Map.lookup bPair (pairGraphObservableReports gr) of
+    -- we have already found observable event differences at this location, so skip the check
+    Just _ -> return gr
+    Nothing ->
+      do res <- doCheckObservables asm bundle preD
+         case res of
+           ObservableCheckEq ->
+             do traceBundle bundle "Observables agree"
+                return gr
+           ObservableCheckError msg ->
+             do traceBundle bundle ("Error checking observables: " ++ msg)
+                return gr
```

Comment:
Should this error be recorded somewhere? Or is that unimportant for some reason?

Reply:
I should be, yes. I think the big question is deciding how severe this class of error is. At first blush, I think these should be recoverable errors (i.e., continue the analysis) but they should definintely be recorded somewhere.  Similar with errors due to widening, etc.